include/sympack/symPACKMatrix2D.hpp

```diff
@@ -1163,11 +1455,37 @@ namespace symPACK{
 
           Int ldsol = tgt_contrib.width();
           Int ldfact = this->width();
-
           if ( this->i == this->j ) {
             bassert(this->blocks().size()==1);
             auto diag_nzval = this->_nzval;
+#ifdef CUDA_MODE
+            if (ldsol*ldfact > symPACK::trsm_limit && symPACK::gpu_solve) {
+            
+                dev_ptr d_diag_nzval = symPACK::gpu_allocator.allocate<T>(ldfact*ldfact);
+                dev_ptr d_nzval = symPACK::gpu_allocator.allocate<T>(ldsol*ldfact);
```

Comment:
These device allocations (and four more similar blocks below) are not checking whether the allocation succeeded. We don't necessarily need a fallback implementation to handle every such allocation failure, but currently the behavior is likely to be a SEGV inside the `upcxx::copy` operation below.

I'd prefer an assertion to provide a clear diagnosis of what went wrong. Eg:

```suggestion
                dev_ptr d_diag_nzval = symPACK::gpu_allocator.allocate<T>(ldfact*ldfact);
                dev_ptr d_nzval = symPACK::gpu_allocator.allocate<T>(ldsol*ldfact);
                UPCXX_ASSERT_ALWAYS(d_diag_nzval && d_nzval,
                  "GPU device segment exhausted, try increasing `-gpu_mem` (see README.md)").
```

If you're feeling fancy you could even factor the message and include the values of `symPACK::gpu_allocator.segment_size()` and `symPACK::gpu_allocator.segment_used()` in the error message, but such queries should be conditional on `#if UPCXX_SPEC_VERSION >= 20230300`

Reply:
Thanks for catching these memory leaks. I will implement and push a fix.

Reply:
Fixed in [97a3e5f](https://github.com/symPACK/symPACK/pull/12/commits/97a3e5f1f6d37fa0e56f7c6290ef8c9722114254)